ceci/pipeline.py

```diff
@@ -84,20 +226,213 @@ def ordered_stages(self, overall_inputs):
+        """Generate the complete command to run a stage.
+
+        This includes any mpirun commands, settings of env vars, and
+        if docker/shifter are used it includes that too.  This depends
+        on the execution settings (nprocess, nthread) as well as the
+        stage itself.
+
+        Parameters
+        ----------
+        stage: PipelineStage
+            The stage to generate the command for
+
+        inputs: dict{str: str}
+            Mapping of inputs to the stage to paths
+
+        config: str
+            Path to stage configuration file for pipeline stages
+
+        outdir: str:
+            Path to directory in which to put outputs
+
+        Returns
+        -------
+        cmd: str
+            Complete command to be executed
+
+        """
+        core = stage.generate_command(inputs, config, outdir, missing_inputs_in_outdir=True)
+        sec = self.stage_execution_config[stage.name]
+        cmd = sec.site.command(core, sec)
+        return cmd
+
+
+
+
+class DryRunPipeline(Pipeline):
+    """A pipeline subclass which just does a dry-run, showing which commands
+    would be executed.
+
+    See the base class for almost all behaviour.
+
+    No additional attributes.
+    """
+    def run(self, overall_inputs, output_dir, log_dir, resume, stages_config):
+        """Dry-run the pipeline and print out all stage command lines
+
+        Some of the arguments here are unused by this sub-class, but are retained
+        for consistency with other pipeline subclasses.
+
+        In case useful for scripting, this command also returns a list of all the
+        commands it prints.
+
+        Parameters
+        ----------
+        overall_inputs: dict{str: str}
+            Any inputs that do not need to be generated by the pipeline but are
+            instead already supplied at the start.  Mapping is from tag -> path.
+
+        output_dir: str
+            Path to directory where command would put outputs.
+        
+        log_dir: str
+            Directory in which to write stage log files. Unused.
+
+        resume: bool
+            Whether to resume the pipeline and avoid executing stages whose
+            outputs already exist.  Unused: all stage commands are printed.
+            TODO: Change this.
+
+        stages_config: str
+            Path to configuration file for the pipeline stages.
+
+        Returns
+        -------
+        cmds: list[str]
+            A list of all the command line invocations.
+
+        """
+        cmds = []
+        for stage in self.ordered_stages(overall_inputs):
             sec = self.stage_execution_config[stage.name]
-            cmd = stage.generate_command(overall_inputs, stages_config, output_dir, sec.nprocess, self.mpi_command)
+            cmd = self.generate_full_command(stage, overall_inputs, stages_config, output_dir)
             print(cmd)
             print()
+            cmds.append(cmd)
+        return cmds
```

Comment:
I've abstracted this now, moving everything into the base class along the lines you've suggested.  It's simplified things nicely - thanks!

I save `run_info` for each pipeline, which is the subclass-specific run data, like you said, and `pipeline_outputs` separately, which contains all the pipeline outputs.  I've put an error status as the return code.

I've also renamed `data_elements` to `pipeline_files` since it was a bit opaque.